endpoints/s3/endpoint.go

```diff
@@ -0,0 +1,263 @@
+
+	return &newEndpoint, nil
+}
+
+func (e *Endpoint) Provider() string {
+	region := ""
+	if e.Client.Options().Region != "" {
+		region = "." + e.Client.Options().Region
+	}
+	baseUrl := "s3" + region + ".amazonaws.com"
+	if e.Client.Options().BaseEndpoint != nil {
+		baseUrl = *e.Client.Options().BaseEndpoint
+	}
+	if e.Client.Options().UsePathStyle {
+		return "s3: " + baseUrl + "/" + e.Bucket
+	}
+	return "s3: " + e.Bucket + "." + baseUrl
+}
+
+func (e *Endpoint) RootDir() string {
+	return e.Bucket + "/"
+}
+
+func (e *Endpoint) FilesStaged(files []any) (bool, error) {
+	staged := true
+	for _, f := range files {
+		descriptor, ok := f.(map[string]any)
+		if !ok {
+			return false, fmt.Errorf("invalid descriptor format")
+		}
+		pathVal, found := descriptor["path"]
+		if !found {
+			return false, fmt.Errorf("descriptor missing 'path' field")
+		}
+		pathStr, ok := pathVal.(string)
+		if !ok {
+			return false, fmt.Errorf("'path' field is not a string")
+		}
+		exists, err := e.fileExists(pathStr)
+		if err != nil {
+			return false, err
+		}
+		if !exists {
+			staged = false
+			break
+		}
+	}
+	return staged, nil
+}
+
+func (e *Endpoint) Transfers() ([]uuid.UUID, error) {
+	ids := make([]uuid.UUID, 0)
+	for id := range e.TransfersMap {
+		ids = append(ids, id)
+	}
+	return ids, nil
+}
+
+func (e *Endpoint) Transfer(dst Endpoint, files []endpoints.FileTransfer) (uuid.UUID, error) {
+	numFiles := len(files)
+	numTransferred := 0
+
+	for _, file := range files {
+		// get the object from the source endpoint
+		buffer, _, err := e.getObject(file.SourcePath)
+		if err != nil {
+			// skip this file
+			continue
+		}
+
+		// put the object into the destination endpoint
+		err = dst.putObject(file.DestinationPath, bytes.NewReader(buffer.Bytes()))
+		if err != nil {
+			// skip this file
+			continue
+		}
+
+		numTransferred++
+	}
+
+	// generate a transfer task ID
+	taskId := uuid.New()
+
+	if numTransferred != numFiles {
+		e.TransfersMap[taskId] = endpoints.TransferStatus{
+			Code:                endpoints.TransferStatusFailed,
+			Message:             fmt.Sprintf("Transferred %d out of %d files", numTransferred, numFiles),
+			NumFiles:            numFiles,
+			NumFilesTransferred: numTransferred,
+			NumFilesSkipped:     numFiles - numTransferred,
+		}
+	} else {
+		e.TransfersMap[taskId] = endpoints.TransferStatus{
+			Code:                endpoints.TransferStatusSucceeded,
+			Message:             "All files transferred successfully",
+			NumFiles:            numFiles,
+			NumFilesTransferred: numTransferred,
+			NumFilesSkipped:     0,
+		}
+	}
+
+	return taskId, nil
+}
+
+func (e *Endpoint) Status(id uuid.UUID) (endpoints.TransferStatus, error) {
+	status, found := e.TransfersMap[id]
+	if found {
+		return status, nil
+	}
+	return endpoints.TransferStatus{}, fmt.Errorf("unknown transfer ID %s", id.String())
+}
+
+func (e *Endpoint) Cancel(id uuid.UUID) error {
+	return fmt.Errorf("transfer cancellation not supported for S3 endpoints")
```

Comment:
We could also add an `endpoints.CancellationNotSupportedError` (though I don't know that it matters just yet)

Reply:
with the asynchronous file transfer, we can now support canceling transfers